src/react/DevicesPage.jsx

```diff
@@ -35,126 +36,150 @@ import SingleCardPage from './ui/SingleCardPage';
+    showNeverConnected = true,
+  } = filters;
+
+  if (!showConnected && device.connected) {
+    return false;
+  }
+  if (!showDisconnected && !device.connected && device.lastConnection) {
+    return false;
+  }
+  if (!showNeverConnected && !device.connected && !device.lastConnection) {
+    return false;
+  }
+  if (deviceId === '') {
+    return true;
+  }
+
+  const aliases = Array.from(device.aliases.values());
+  return (
+    device.id.includes(deviceId) || aliases.filter((alias) => alias.includes(deviceId)).length > 0
+  );
+};
+
 export default ({ astarte, history }) => {
   const [phase, setPhase] = useState('loading');
-  const [totalDevices, setTotalDevices] = useState(null);
   const [activePage, setActivePage] = useState(0);
-  const [maxPage, setMaxPage] = useState(0);
-  const [cachedPages, setCachedPages] = useState([]);
-  const pageAlerts = useAlerts();
+  const [deviceList, setDeviceList] = useState([]);
+  const [requestToken, setRequestToken] = useState(false);
+  const [showSidebar, setShowSidebar] = useState(true);
+  const [filters, setFilters] = useState({});
 
-  const handleDevicesRequest = (pageIndex, response) => {
-    const token = new URLSearchParams(response.links.next).get('from_token');
-    const deviceList = response.data.map((value) => Device.fromObject(value));
-    cachedPages[pageIndex] = {
-      devices: deviceList,
-      token,
-    };
-    setCachedPages([...cachedPages]);
-    if (pageIndex === activePage) {
-      setPhase('ok');
-    }
-    const pagesToCache = activePage + MAX_SHOWN_PAGES + 1;
-    if (pageIndex < pagesToCache) {
-      // eslint-disable-next-line no-use-before-define
-      cachePage(pageIndex + 1);
+  const pageAlerts = useAlerts();
+  const pagedDevices = useMemo(() => {
+    const pageList = [];
+    const devices = deviceList.filter((device) => matchFilters(device, filters));
+    while (devices.length > 0) {
+      pageList.push(devices.splice(0, DEVICES_PER_PAGE));
     }
-    return null;
-  };
+    return pageList;
+  }, [deviceList, filters]);
 
-  const cachePage = (pageIndex) => {
-    if (pageIndex === 0) {
-      astarte
-        .getDevices({
-          details: true,
-          limit: DEVICES_PER_PAGE,
-        })
-        .then((response) => handleDevicesRequest(pageIndex, response))
-        .catch((err) =>
-          pageAlerts.showError(`Couldn't retrieve the device list from Astarte: ${err.message}`),
-        );
-    } else {
-      const { token } = cachedPages[pageIndex - 1];
-      if (token) {
-        astarte
-          .getDevices({
-            details: true,
-            from: token,
-            limit: DEVICES_PER_PAGE,
-          })
-          .then((response) => handleDevicesRequest(pageIndex, response))
-          .catch((err) =>
-            pageAlerts.showError(`Couldn't retrieve the device list from Astarte: ${err.message}`),
-          );
-      }
+  const loadMoreDevices = async (fromToken, loadAllDevices = false) =>
```

Comment:
I think we should give a more meaningful name to `fromToken` and `nextToken`. Maybe `nextPageLink`?
I think it would also play fine with
```
<TablePagination
    ...
    canLoadMorePages={!!nextPageLink}
    ...
/>
```

Reply:
This is not a link, it's a token. A string of numbers that will tell Astarte where to start from when retrieving the device list

Reply:
That's right. However I actually wanted to point out that having a `page` reference can aid in understanding the variable's meaning.
`nextPageToken`?

Reply:
Let's keep as it is for now, ideally the Astarte client should provide a paginator (or some iterator) for the device list.
This way we can hide the token logic and make the code more readable.

Reply:
I agree